hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/lock/OzoneManagerLock.java

```diff
@@ -172,7 +176,16 @@ private boolean lock(Resource resource, String resourceName,
       LOG.error(errorMessage);
       throw new RuntimeException(errorMessage);
     } else {
+      long startWaitingTimeNanos = Time.monotonicNowNanos();
+      /**
+       *  holdCount helps in metric updation only once in case of reentrant
+       *  locks.
+       */
+      int holdCount = manager.getActiveLockCount(resourceName);
       lockFn.accept(resourceName);
+      if (holdCount == 0) {
+        updateReadLockMetrics(resource, lockType, startWaitingTimeNanos);
+      }
```

Comment:
We should consider the case that another thread is holding the lock when this thread tries to acquire it.  My understanding is that `holdCount` will be `>= 1`, so we skip updating lock metrics.

Similarly, `unlock` only updates held time for the last thread that releases the lock (read lock can be held at the same time by multiple threads).

Since lock acquire/release and count increase/decrease are two separate, non-atomic steps, I don't think we can rely on `holdCount` for this.

Reply:
Here's a test case to reproduce the multiple readers problem I mentioned:

```java
  @Test
  public void testReadLockConcurrentStats() throws InterruptedException {
    // GIVEN
    final int threadCount = 5;
    OzoneManagerLock.Resource resource = OzoneManagerLock.Resource.VOLUME_LOCK;
    String[] resourceName = generateResourceName(resource);
    OzoneManagerLock lock = new OzoneManagerLock(new OzoneConfiguration());
    Thread[] threads = new Thread[threadCount];
    for (int i = 0; i < threads.length; i++) {
      threads[i] = new Thread(() -> {
        lock.acquireReadLock(resource, resourceName);
        try {
          Thread.sleep(1000);
        } catch (InterruptedException e) {
          e.printStackTrace();
        }
        lock.releaseReadLock(resource, resourceName);
      });
      threads[i].start();
    }
    for (Thread t : threads) {
      t.join();
    }

    // WHEN
    String heldStat = lock.getReadLockHeldTimeMsStat();

    // THEN
    Assert.assertTrue("Expected " + threadCount + " samples in " + heldStat,
        heldStat.contains("Samples = " + threadCount));
  }
```

Result:

```
[ERROR] Failures:
[ERROR]   TestOzoneManagerLock.testReadLockConcurrentStats:389 Expected 5 samples in Samples = 1  Min = 1.26694054E8  Mean = 1.26694054E8  Std Dev = 0.0  Max = 1.26694054E8
```

Only one of the 5 threads that acquire the read lock are reflected by the metrics stat.

Similar test case can be constructed for 1 writer + N readers.

Reply:
Sure @adoroszlai, thank you so much for pointing out - will be looking into the same!

Reply:
Hi @adoroszlai, please find the changes addressing the multiple readers metrics updation problem, thanks